### ISSUE TITLE
Costumes and images for sprites and backgrounds disappear after reopening saved project

### ISSUE THREAD

**Issue by mkGit-1** [jmoenig/Snap#2535]
Students will save a project to the cloud with their sprites and costumes working perfectly both imported costumes and native ones, and when they log in every image they saved will have disappeared including backgrounds. This only happens intermittently so I have no idea what exactly is needed to reproduce it. Nothing else is deleted not sprites, names, scripts variables just the images they saved with the project. The sprites, scripts, and backgrounds will all be there just with blank spaces instead of images.

**Comment by bromagosa** [comment]
Are you by any chance using Chrome v77?

**Comment by mkGit-1** [comment]
> Are you by any chance using Chrome v77?

I’m not sure the exact version but we are using chrome. However we opened the projects in Microsoft edge and they didn’t appear either but we didn’t try saving in Edge

**Comment by bromagosa** [comment]
If you have the opportunity, please check the version of the Chrome browser you used to save the projects. If it's v77, then we're pretty sure a known bug in that version is causing this.

**Comment by mkGit-1** [comment]
> If you have the opportunity, please check the version of the Chrome browser you used to save the projects. If it's v77, then we're pretty sure a known bug in that version is causing this.

No we are using v79

**Comment by bromagosa** [comment]
@jmoenig I've a bit disconnected from this for a bit. Are the v79 issues affecting canvas operations too?

@mkGit-1, can you try saving in Firefox and see if the issue persists?

**Comment by jmoenig** [comment]
@bromagosa yes, unfortunately v79 has tremendous issues with Canvas, once again. This time the problem is not speed (although the issues of v78 persist, but those were better than those v76 and v77) but - once again - stability. v79 frequently white-outs, and then it seems that all of Canvas is broken and doesn't render at all or not correctly anymore. I haven't done any conclusive research on this, but it could be that to saving a project after such rendering deficiencies occur leads to loss of image data (because Canvas is broken in that state). This might explain why costumes get lost when Chrome is already in a broken state. But, at this time it's just a speculation.

Some users report that disabling the GPU in chrome://flags helps, but makes things slower otherwise.

I'm working on a strategy to fundamentally redesign Morphic in 3 phases so we have more control over its memory consumption. But I'm really in the dark about what's going on, especially since Chrome doesn't bother to let us know what's wrong. There are never any error messages or warnings when this happens. I'm also not sure how much of this can be attributed to our support for higher devicePixelRatios ("retina support"). While we're doing the "official right way" this way is obviously so kludgy that I wouldn't be surprised if Chrome devs didn't remember to think about it when they change things in their code.

I've gotten the first phase of the Morphic redesign ready over the last week and will start committing it either later today or early next week. It will slightly reduce Snap's memory footprint and accelerate loading Snap projects in Chrome by 20%-45%. It will also slightly reduce the white-outs, but not be a real solution yet. So it might not make sense to release it as a minor version just yet.

In a second phase I'm planning to change Morphic in a more profound way and move away from always caching every Morph's surface property in its own canvas to a more differentiated approach where we can cache those Morphs that don't need frequent redraws and don't cache those that we're constantly redrawing anyway. I've also designed a way that will let us reduce the total number of Canvas elements involved in all of Snap to just 2. This will likely slow down Snap! dramatically, but it will also make it super stable again. I'm hoping that we'll be able to find a sweet spot between speed and stability this way. Once we've found such an optimum I'll be able to sleep better. This is currently my top priority, because I don't trust Google to fix Chrome permanently. They've shown several times now that they have no clue whatsoever about the effects of what they believe might be subtle changes, and don't react to bug reports in a timely and responsive manner, so we better address this ourselves. We might even end up checking the browser manufacturer and special casing Chromium, because Firefox and Safari don't have these issues (Safari, of course, most of the time has its own issues).

In the longer run - the third phase - I'm planning to move Morphic away from being a timesharing multiplexer and towards a multi-core architecture. By using hardware parallelization to run atomic process steps in parallel we should be able to dramatically scale up what Morphic and Snap! can accomplish per display cycle, and the need for extensive caching might subside. The technology for this is currently still experimental and only partially supported by some browsers: OffscreenCanvas and WebWorkers. Once this becomes a standard and is supported by all major browsers this could be a significant "quantum" kinda leap for Snap!. And it even looks like we could start exploring this very soon.

As an alternative I've looked into SVG to find out whether moving away from Canvas to an all-vector Morphic based on SVG might be feasible and a good idea. Feasibility doesn't look like a problem, but I'm not sure about performance gains, especially keeping in mind that Dan Ingalls tried just that in the first iterations of his LivelyKernel, only to ditch it for being slow and clunky.

**Comment by bromagosa** [comment]
Thanks for the detailed explanation, Jens. I would guess that SVG transformations are not only going to be way slower, but also most probably quite changing from browser to browser, so it seems we'll need to stick to canvas and find a way around these nasty Chrome bugs.

We should maybe also start recommending Firefox as a first choice instead of Chrome in the [requirements](https://snap.berkeley.edu/requirements) page...

**Comment by jmoenig** [comment]
Oh sure, let's update the requirements page! But I think we really cannot afford not to recommend Chrome, but we can add that we're *also* recommending Firefox :-) How about for both browsers we write:

    Firefox / Chrome are currently the recommended browsers for Snap!

Does that make sense?

**Comment by brianharvey** [comment]
On 1/3/20 2:47 AM, Jens Mönig wrote:
> especially keeping in mind that Dan Ingalls tried just that in the first iterations of his LivelyKernel, only to ditch it for being slow and clunky.

Of course that was in the early days of SVG support; who knows what it's 
like today.

**Comment by brianharvey** [comment]
> But I think we really cannot afford not to recommend Chrome,
We can't afford not to _support_ Chrome, but I don't see why we have to recommend it.

**Comment by jmoenig** [comment]
yeah, SVG has not gotten much faster. I've done quite a few experiments, because it seemed like a promising strategy to explore, but I was disappointed. Bitmaps appear to be much more performant.
